### DISCUSSION TITLE
Fixing two way binding for react signals

### DISCUSSION THREAD

**Discussion by MahmoodKhalil57** [preactjs/signals#518] [Ideas]
The preact is an awesome library that solved a big issue I had with react, lack of signals. This library is single handedly convincing me to move from svelte to react. While playing around with it I found that I had an issue with two way binding an input field.

First test:
```tsx
<input type="text" value={text.value} onChange={(e) => { text.value = e.currentTarget.value }} />
```
The above did not work because even though the value of the signal changed successfully, it doesnt reflect back to the field and update it to the new value, ie. the value of the input field is stuck on the original text.value

### The Problem:
Second test:
```tsx
<input type="text" value={text} onChange={(e) => { text.value = e.currentTarget.value }} />
```
The above worked successfully! but it is throwing the following misleading error

`Type 'Signal<string>' is not assignable to type 'string | number | readonly string[] | undefined'.
  Type 'Signal<string>' is missing the following properties from type 'readonly string[]': length, concat, join, slice, and 19 more.ts(2322)
index.d.ts(3325, 9): The expected type comes from property 'value' which is declared here on type 'DetailedHTMLProps<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>'`

Im calling it misleading because it acually works, and it successfully does the two way binding exactly as I expect, where the value of the input field and the signal updates when I type something in the input box.

### The Solution:
Solution 1:
First thing that came to my mind was to fix the types, since im not really a typescript or react expert, it took me a while to reach this solution but it works fine for me for now. I added an index.d.ts file in src and added the following code to it, it is a bit ugly, but it fixes the type issue, allowing Signal<string> to be used as a value in the input field
```ts
declare const identifier: unique symbol;

declare class Signal<T = any> {
  constructor(value?: T);
  subscribe(fn: (value: T) => void): () => void;
  valueOf(): T;
  toString(): string;
  toJSON(): T;
  peek(): T;
  brand: typeof identifier;
  get value(): T;
  set value(value: T);
}

interface ExtendedInputHTMLAttributes<T> extends React.InputHTMLAttributes<T> {
  value?: string | readonly string[] | number | Signal<string> | Signal<readonly string[]> | Signal<number> | undefined;
}

declare module JSX {
  interface IntrinsicElements {
    input: React.DetailedHTMLProps<ExtendedInputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;
  }
}
```

This solution was kindof ugly to me since I was making an assertion about the types that might be false, again, im not an expert on this so it felt wrong to do that.

Solution 2:
After some googling, I found a post that recommends creating a custom directive, so I did. In it I can infer the type of the value inside the signal, and use it to assert the type of the value, I believe I can safely do that since this is what Signals do in the end (would love if someone better than me could verify this)
```ts
export const bindInput = <T extends Signal>(signal: T): React.InputHTMLAttributes<HTMLInputElement>  => {
  type SignalType = T extends Signal<infer U> ? U : never
  return {
    value: signal as SignalType,
    onInput: (event) => (signal.value = event.currentTarget.value)
  }
}
```

it can then be used simply like so
```tsx
<input type="text" {...bindInput(text)} />
```

I would have loved to make this into a pull request, but Im not really sure where I should make these changes in the project. Plus I dont think either of the solutions are perfect so maybe it is better for someone from the preact team might take inspiration from this and implement something even better.

Thank you for you hard work!

**Comment by XantreDev** [comment]
Now preact signals works by babel plugin which checks `.value` access inside your function components. Props unwrapping for now shouldn't work, because `signals-react` is not patching react anymore
```tsx
const text = signal('')
// no `.value` access inside the component - will not be tracked
const A = () => <input type="text" {...bindInput(text)} />
```

```tsx
const text = signal('')
/**
* @useSignals
* not sure that it will be correctly tracked, because signal probably unwraps while setting value to actual DOM
**/
const A = () => <input type="text" {...bindInput(text)} />
```

**Reply by MahmoodKhalil57** [reply]
sorry this is a bit too technical for me, do you mean the code I provided shouldnt work because unwrapped props arnt tracked?

I did test the code I provided before making this post, here is a preview
https://react-test-inky-iota.vercel.app/
https://github.com/MahmoodKhalil57/react-test

**Reply by XantreDev** [reply]
I actually surprised that this is working at all, because you have no babel plugin setup: https://github.com/preactjs/signals/tree/main/packages/react#babel-transform

**Reply by MahmoodKhalil57** [reply]
Im having mixed feelings about that comment lol. I just installed react via vite, and installed the preact signals library and it worked great, didnt need to do alot to make it work exactly as I needed. Hope this helped!

**Reply by XantreDev** [reply]
Just try to use 
```tsx
const count = signal(0)
setInterval(() => count.value++, 1000)
function Counter() {
	return <p>Value: {count.value}</p>;
}
```

**Reply by MahmoodKhalil57** [reply]
Okay I think Im starting to understand what you mean, but surprisingly it still works
I changed it to this and it worked fine
```tsx
const count = signal(0)
setInterval(() => count.value++, 1000)
function Counter() {
	return <p>Value: {count}</p>; // used the signal instead of value
}
```

https://react-test-inky-iota.vercel.app/
https://github.com/MahmoodKhalil57/react-test/blob/main/src/SignalCounter.tsx

**Reply by XantreDev** [reply]
This will work, because signals are valid jsx elements, but components will not track signal changes

**Reply by MahmoodKhalil57** [reply]
As in the entire component wont rerender, only the html bit would? This is exactly what I want. If im misunderstanding this can you lead me to a doc that I can read into?

**Reply by rschristian** [reply]
The problem that @XantreGodlike is getting at is that `signal.value` won't be tracked in components -- this might be okay if you're only using it directly as a text node, but the second you need to use the value in an expression (like `<h1>{signal.value ? 'True' : 'False'}</h1>`) it'd break (without the Babel plugin). This might not be an issue for your demo case but we certainly don't recommend using it without. That's a pretty big footgun that you'd get confused about later.

As for docs, the React integration ReadMe is all there is for React.

**Reply by MahmoodKhalil57** [reply]
+1

**Comment by rschristian** [comment]
I don't think patching all of React's types to accept signals is within the scope. It'd be way too much maintenance.

You certainly could create a directive if you wished, but that'd be something for user code. Here's a [Preact implementation](https://gist.github.com/marvinhagemeister/daada86f82466dff81d48840c8d5f27b) of a directive.

To be clear, a directive would be something like this: `<input bind:value={signal} />`. What you have isn't a directive, just a utility function to return two properties. If that works for you though, then great.

**Comment by MahmoodKhalil57** [comment]
Note: I was not using the `@preact/signals-react-transform` babel plugin, my experience stated above will probably be different than the classic experience.

**Comment by developit** [comment]
Your first code snippet is the correct one. You just need to add the hook to make your component reactive:

```js
import {useSignals} from '@preact/signals-react/runtime';

function MyComponent() {
  useSignals(); // enable automatic re-rendering

  return (
    <input type="text" value={text.value} onChange={(e) => { text.value = e.currentTarget.value }} />
  );
}
```

Alternatively, you can use the Babel plugin that does this for you.
